File: app/src/main/java/com/esorokin/lantector/ui/fragment/BaseFragment.java
package com.esorokin.lantector.ui.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.esorokin.lantector.ui.plugins.base.CompositionPlugin;
import com.esorokin.lantector.ui.plugins.base.CompositionPluginDelegate;

public class BaseFragment extends MvpAppCompatFragment {
	private final CompositionPlugin compositionPlugin = new CompositionPluginDelegate();

	/**
	 * Attach your plugins here.
	 */
	protected void initPlugins() {
		//override
	}

	protected CompositionPlugin compositionPlugin() {
		return compositionPlugin;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		initPlugins();
		compositionPlugin.onCreate();
	}

	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		compositionPlugin.onViewCreated(view);
	}

	@Override
	public void onStart() {
		super.onStart();
		compositionPlugin.onStart();
	}

	@Override
	public void onResume() {
		super.onResume();
		compositionPlugin.onResume();
	}

	@Override
	public void onPause() {
		super.onPause();
		compositionPlugin.onPause();
	}

	@Override
	public void onStop() {
		super.onStop();
		compositionPlugin.onStop();
	}


	@Override
	public void onDestroy() {
		super.onDestroy();
		compositionPlugin.onDestroy();
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		compositionPlugin.onActivityResult(requestCode, resultCode, data);
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/fragment/BaseFragment.kt
package com.esorokin.lantector.ui.fragment

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View
import com.esorokin.lantector.ui.plugin.MoxyFragmentPlugin
import com.esorokin.lantector.ui.plugin.base.CompositionPluginDelegate

abstract class BaseFragment : Fragment() {
    protected val compositionPlugin = CompositionPluginDelegate()

    /**
     * Attach your plugins here.
     */
    protected open fun initPlugins() {
        compositionPlugin.attach(MoxyFragmentPlugin(this))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initPlugins()
        compositionPlugin.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        compositionPlugin.onViewCreated(view!!)
    }

    override fun onStart() {
        super.onStart()
        compositionPlugin.onStart()
    }

    override fun onResume() {
        super.onResume()
        compositionPlugin.onResume()
    }

    override fun onPause() {
        super.onPause()
        compositionPlugin.onPause()
    }

    override fun onStop() {
        super.onStop()
        compositionPlugin.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        compositionPlugin.onDestroy()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        compositionPlugin.onActivityResult(requestCode, resultCode, data)
    }
}
